File: aoc2023/haskell-solutions/src/Day16.hs
module Day16 (day16) where

import Control.Parallel.Strategies (parMap, rpar)
import Coords (Coord(..))
import qualified Coords as C
import qualified Data.Set as Set
import Data.Vector (Vector)
import qualified Data.Vector as V
import Util

day16 :: IO (Int, Int)
day16 = do
  inp <- parse
  let xmax = V.length inp         - 1
      ymax = V.length (inp V.! 0) - 1

      one = [(C 0 0, R)]
      two = concatMap (\x -> [(C x 0, D), (C x ymax, U)]) [0..ymax]
         ++ concatMap (\y -> [(C 0 y, R), (C xmax y, L)]) [0..xmax]

      solve = maximum . parMap rpar (Set.size . Set.fromList . map fst . simulate inp)

  pure (solve one, solve two)

parse :: IO (Vector (Vector Char))
parse = V.fromList . map V.fromList . transpose . lines <$> readFile' "../inputs/day16.txt"

reflectFwd :: Coord -> Dim4 -> (Coord, Dim4)
reflectFwd beam = \case
  R -> (C.above beam, U)
  U -> (C.right beam, R)
  D -> (C.left  beam, L)
  L -> (C.below beam, D)

reflectBck :: Coord -> Dim4 -> (Coord, Dim4)
reflectBck beam = \case
  R -> (C.below beam, D)
  D -> (C.right beam, R)
  U -> (C.left  beam, L)
  L -> (C.above beam, U)

simulate :: Vector (Vector Char) -> (Coord, Dim4) -> [(Coord, Dim4)]
simulate grid start = bfs [start] \(beam, dir) ->
  filter (isJust . C.ix grid . fst) $
    case C.ix grid beam of
      Just '.'  -> [(C.move beam dir, dir)]
      Just '/'  -> [reflectFwd beam dir]
      Just '\\' -> [reflectBck beam dir]
      Just '-'  -> if dir `elem` [U, D]
                   then [(C.left beam, L), (C.right beam, R)]
                   else [(C.move beam dir, dir)]
      Just '|'  -> if dir `elem` [L, R]
                   then [(C.above beam, U), (C.below beam, D)]
                   else [(C.move beam dir, dir)]
      _ -> []

File: aoc2023/clojure-solutions/src/clojure_solutions/day16.clj
(ns clojure-solutions.day16
  (:require [clojure.string :as str]
            [clojure-aoc-util.util :as aoc]
            [clojure-aoc-util.coords :as c]))

(defn- parse []
  (aoc/transpose (str/split-lines (slurp "../inputs/day16.txt"))))

(defn- simulate [grid start]
  (aoc/flood-fill
   start
   (fn [[beam dir] seen]
     (filter #(and (aoc/not-in? seen %)
                   (aoc/mat-ix grid (first %)))
             (case (aoc/mat-ix grid beam)
               \.  [[(c/move dir beam) dir]]
               \/  [(case dir
                      :E [(c/above beam) :N]
                      :N [(c/right beam) :E]
                      :S [(c/left  beam) :W]
                      :W [(c/below beam) :S])]
               \\  [(case dir
                      :E [(c/below beam) :S]
                      :S [(c/right beam) :E]
                      :N [(c/left  beam) :W]
                      :W [(c/above beam) :N])]
               \-  (if (contains? #{:N :S} dir)
                     [[(c/left beam) :W], [(c/right beam) :E]]
                     [[(c/move dir beam) dir]])
               \|  (if (contains? #{:E :W} dir)
                     [[(c/above beam) :N], [(c/below beam) :S]]
                     [[(c/move dir beam) dir]]))))))

(defn day16 [p]
  (let [grid (parse)
        xmax (count grid)
        ymax (count (first grid))
        starts (case p
                 :one [[[0 0] :E]]
                 :two (concat
                       (mapcat (fn [x] [[[x 0] :S] [[x (dec ymax)] :N]])
                               (range xmax))
                       (mapcat (fn [y] [[[0 y] :E] [[(dec xmax) y] :W]])
                               (range ymax))))]
    (->> starts
         (pmap #(into #{} (map first) (simulate grid %)))
         (map count)
         (apply max))))
